src/Types/TypeRegistry.php

```diff
@@ -26,6 +59,22 @@ public function __construct(array $instances = [])
         $this->instances = $instances;
     }
 
+    public static function getInstance(): self
+    {
+        return self::$singleton ??= self::createSingleton();
```

Comment:
IIRC, earlier we discussed that the type registry should become scoped to the wrapper connection at some point. Now looks like the right time to introduce `Connection::getTypeRegistry(): TypeRegistry` and recommend using it.

Promoting the usage of `TypeRegistry::getInstance()` feels like back in pre-`Zend_Registry` times 🤮

Reply:
Also, speaking of the registries. At this point, it sounds like `TypeRegistry` and `getTypeRegistry()` are not the best names for the API.

At the time of the introduction of the type registry (https://github.com/doctrine/dbal/pull/3354), technically, it was a set: each type could be (registered) in or not (registered) in the set/registry. Now it's a map of userland type names to the objects implementing the types.

I believe we should name the new method `getTypeMapping()` or `getTypeMap()` and later rename the registry class accordingly.

Reply:
If I introduce such a property in the wrapper Connection, and then I start using it in that class, the test suite breaks because calls to `Type::addType()` won't have an effect on the instance stored in the connection object. I see an ugly solution to that issue: expose the existence of the singleton to `TypeRegistry` with a public internal method, and, then, in `TypeRegistry`, fallback to that singleton if it is defined. That way, there can be support for cases where part of your application has migrated to scope typed registry, while another part continues using the singleton. Do you see a more simple solution?

Reply:
Also, I don't know with what this should be replaced: https://github.com/doctrine/dbal/blob/8532d4942c2b1565ad27a211966ede334d68d5f1/src/Schema/Table.php#L349-L356

Reply:
I don't think this workaround is desired because it masks an API design issue: we want the type mapping to be scoped to a connection but we want schema objects to be unaware of the connections but aware of the type mapping.

We could do some prototyping and see what it will take to make the connection-scoped type mapping accessible to schema objects, e.g. add an argument of `TypeRegistry` to all schema object methods that depend on it (e.g. to `Table::addColumn()`). Alternatively, instead of accepting a string, those methods could accept the type instance.

We can try prototyping without taking BC into account just to see what the resulting API changes will look like. We can thing of the BC and the upgrade path later.

Reply:
Ah I remember brainstorming a bit about this already: https://github.com/doctrine/dbal/pull/5034
I see I mention there that API inside `AbstractPlatform` that rely on the type registry have been deprecated and removed but… that's not the case. Not sure what I thought, but maybe these APIs _should_ be deprecated. Their purpose is unclear.

Reply:
Are you talking about `registerDoctrineTypeMapping()` and `getDoctrineTypeMapping()`? In my understanding, their purpose of mapping DBAL types to platform types but using the terminology from https://github.com/doctrine/dbal/pull/5406#discussion_r878495537, those might be only the "DBAL/ORM-level types". E.g. you cannot add a custom `JSON` type if a platform doesn't support it, you can only emulate it in the DBAL. At this point, the platform doesn't need to be aware of this type.

So, their purpose is clear but it's a design mistake and those methods may need to be eliminated to enable further type system improvements.

Reply:
> You cannot add a custom "platform-level type" (e.g. JSON) if the platform doesn't support it, you can only emulate it in the DBAL.

I think you might be able to, if you register that type in the type registry first, and have your type implement `getMappedDatabaseTypes`

I think that's what's done here: https://github.com/doctrine/dbal/blob/3.3.x/tests/Functional/Schema/MySQL/PointType.php#L1-L35

It's used in this test: https://github.com/doctrine/dbal/blob/3.3.x/tests/Functional/Schema/MySQLSchemaManagerTest.php

So now my understanding is:
1. the type registry has all types
2. the platform restricts itself to a subset of those, so that you can only use some of those in `SchemaManager::_getPortableTableColumnDefinition()`
3. there is a check for extra types through the other convoluted mechanism described above.

If we are to scope the type registry to a connection, since a connection has exactly one platform, it might make sense to take that into account and have type registries that restrict themselves to types supported by the platform. We would have to support adding custom types to those registries, hopefully in a more straightforward manner.

Reply:
> I think you might be able to […]

You're right.

Now the question is: how do we integrate the registry into the lifecycle of the connection and platforms? I.e. what initializes what, at what point in time, and what dependencies do components pass to each other?

> [...] have your type implement `getMappedDatabaseTypes`

This looks like a quite rudimentary API. No surprise that the standard types don't use it. Instead, this mapping is hard-coded in the platform classes.

Reply:
I think the `TypeRegistry` might best be platform-scoped (instead of connection-scoped): the schema managers know about the platform, and if the type registry is now the list of types _supported by the platform_ and not a global list of types like before, then it needs information from the platform to be initialized (the list contained in `initializeDoctrineTypeMappings`)

I think the `TypeRegistry` should be initialized by the platform, and I'm undecided on whether platform should expose the type registry to the outer world, or rather proxy methods for it.

About initialization, right now the map in the platform from platform types to portable types is created lazily, and so is the type registry. I'm undecided on whether we should perpetuate that.

On another note, I noticed that a lot of test in this PR switch from being aware of `Type` to being aware of `TypeRegistry`, when they could rely on `Connection::convertToPHPValue()` and `Connection::convertToDatabaseValue`. I'm going to work on making that switch in the test suite, it should make subsequent PRs smaller. Let me know if this is a bad idea.

Reply:
> I think the `TypeRegistry` might best be platform-scoped (instead of connection-scoped)

In this case, it would have nothing to do with user-defined DBAL-level types (e.g. `Rot13`) which are the majority and are platform-independent. We're again circling back towards the problem that the current `Type` API describes two different concepts, and this combination doesn't fit into any of the DBAL layers because these concepts belong to different layers.

> I think the `TypeRegistry` should be initialized by the platform

The platform is instantiated after a physical connection to the database is established. How will the code work before connecting? Will any interaction with the type registry force connection?

Reply:
> The platform is instantiated after a physical connection to the database is established. How will the code work before connecting? Will any interaction with the type registry force connection?

Ah I didn't have that constraint in mind, and I suppose it's not really acceptable to be unable to use some of APIs unless we establish a physical connection prior to that, like for instance `Table::addColumn`.

That might explain why there is this current separation between a global map of portable types in `TypeRegistry`, and a map of supported types keyed by RDBMS-defined types in `AbstractPlatform`.

> We're again circling back towards the problem that the current Type API describes two different concepts, and this combination doesn't fit into any of the DBAL layers because these concepts belong to different layers.

Ok, shall we try to find new names for the 2 concepts / responsibilities and then separate them? After reading what you wrote in https://github.com/doctrine/dbal/pull/5406#discussion_r878495537 , I feel like there should be 2 (or maybe 3 if we split the first one) interfaces with these methods in them

The first one seems to be about changing values before writing to or after reading from the database. I think we should avoid having "type" in the name of this one since it's more about performing operations on values, sometimes even losing information in the process, like with the `UpperCaseStringType`.

```php
interface ValueConverter
{
    public function convertToDatabaseValue(mixed $value, AbstractPlatform $platform);
    public function convertToPHPValue(mixed $value, AbstractPlatform $platform);
    public function convertToDatabaseValueSQL($sqlExpr, AbstractPlatform $platform);
    public function convertToPHPValueSQL($sqlExpr, $platform);
}
```

The second one is about recognizing types that may be named differently on different platforms, but should be considered the same and will be used interchangeably when switching platforms. This part is close to the `doctrineTypeMapping`, for instance for `DB2`, the DBAL type called `text` maps  to `clob`, and `datetime` maps to `timestamp`

Here I think it is fine to keep `Type` in the name

```php
interface Type
{
   abstract public function getSQLDeclaration(array $column, AbstractPlatform $platform);
}
```



We could then have a default value converter for each type (for types that actually need one). It looks like the deprecated `ArrayType` and `ObjectType` could use the same value converter, and so would `BlobType` and `BinaryType`, and many more I suspect.

Apologies if all this is already obvious to you, I'm just trying to wrap my head around all this.

Reply:
Before we get too deep into discussing the above (which I don't see any problems with), let's revisit our conclusions in  https://github.com/doctrine/dbal/pull/4930#issuecomment-952604506.

The primary goal of the DBAL is to provide a database abstraction for writing portable applications. This is why we have the "integer" type mapped to `INTEGER` on some platforms, `NUMBER` on others, and so on.

What is the role of a value converter in building a database abstraction? The user-defined types that are effectively converters belong more to the ORM (if at all) rather than the DBAL.

I'd like to challenge my own statement from the above thread:

>  I know that the built-in types are used w/o the ORM, e.g. the date types.

And ask "so what?" We might do a mental exercise and see what would the DBAL lose in terms of its functionality if value conversion wasn't part of the `Type` API? I wouldn't be surprised if it brings 20% of value by carrying 80% of the complexity.

Reply:
The 2 first methods of the value conversion interface above have proxy methods in `Connection`, so it would still be possible to use them indirectly as long as there is a connection object laying around. The loss of functionality would be more about the 2 other methods I suppose. I don't think these 2 methods belong to the ORM though, should they? They seem to be about storage (for instance, I think they could be used for efficiently storing a uuid in MySQL).

Reply:
> The loss of functionality would be more about the 2 other methods I suppose. I don't think these 2 methods belong to the ORM though, should they?

I don't know where they belong since the standard DBAL types don't use them, and the test examples in the ORM look made up — they both cause a loss of information being stored. Are there real use cases where they would be needed?

I'm very skeptical about maintaining the API that the DBAL itself doesn't use because there's no clear understanding of what it is and what it is not for.

Reply:
Note that user-defined types might also be mapped to some types supported by the platform that the project uses (but not available in DBAL itself due to not being available cross-platform). It is useful for a project relying on Postgresql to map some Postgresql types to DBAL types (making them usable in the same way than other types of columns when using DBAL and the ORM). So this mapping of platform types to DBAL types still benefits from being extensible.

@greg0ire the conversion to SQL (altering the SQL itself) is often tied to the underlying SQL declaration (often being relative here, as most types have a no-op implementation for those methods as this is only useful for more complex types).

And you also forgot `Type::getBindingType`, which is kinda related to them as well.

Reply:
> Are there real use cases where they would be needed?

I think if you wanted to store a UUID as binary in mysql, it would feel safer to use native functions like `UUID_TO_BIN`. That's not an approach used in https://github.com/ramsey/uuid-doctrine/blob/main/src/UuidBinaryType.php though, probably because these functions are too recent, as documented in the README: https://github.com/ramsey/uuid-doctrine/blob/08e16b0dbea8ade7d03b0b377fe37942628eaae6/README.md#working-with-binary-identifiers

Reply:
> And you also forgot `Type::getBindingType`, which is kinda related to them as well.

As for this method specifically, it should be something like `Type::bind($stmt, $param, $value)`. The type should own the logic of binding a value of the type to a statement. Otherwise, the wrapper connection has to deal with three types of statement parameters: the driver-level binding type (int), the type name (string) and the type instance.

Reply:
@morozov do you mean `Type::bind()` should also encapsulate the logic for the 2 first types (driver-level binding type and type name)? If yes, it must be missing an argument. Also, should it be a static method then?

Also, `Connection::quote()` should have a `Type` equivalent if we are going that way:
https://github.com/doctrine/dbal/blob/de42378d0e521fd45d375020472caf37a003d289/src/Connection.php#L803-L818

Or should that method be deprecated? The comments on it suggest it should.

Reply:
It should translate the binding of a value of type X to the driver-level type. E.g. binding an `UserId` object to the statement by extracting the integer value and binding it as an integer. What argument does it missing?

Why should it be a static method? The type object may be responsible for binding values of its own type. The only problem with this is that the value type will remain mixed, so we could think of something better.

The quote method should be eventually deprecated but it's widely used in the ORM. I don't think it's used for quoting anything than strings, so we probably don't need to worry about quoting arbitrary types (it doesn't make much sense in any event).

Reply:
> What argument does it missing?

The type argument I think.

`$stmt` is a driver statement, param an int or a string, with the placeholder number or name, and value is the value of the parameter.

If the method is not static, then we have a `Type` object: `$this`, so we don't need that extra argument for the case of a type instance. In case we have a string, we can look it up the type registry and also get a type instance, but I don't know what should be done for the case of a driver-level binding type.

Reply:
Sorry, that would have been a type instance method, not a static method.

Reply:
Yes, but let's say the type is `1`… if `bind()` is to encapsulate using this as a driver-level binding type (which is an `int`, right?), then how do I pass that `1` to that method? Through which argument?

Reply:
> Yes, but let's say the type is `1`…

How can a type be `1`? It's an object. That's what I mean:
```php
class IntegerType {
    public function bind(Statement $statemet, int|string $param, mixed $value): void {
        $statement->bind($param, $value, ParameterType::INTEGER);
    }
}
```
The logic here is the same as in `Type::getBindingType()` which triggered this discussion with the only difference that the consumer of the `Type` API no longer asks: "what parameter type should I use to bind values of this type?". Instead, it tells: "bind this for me".

Reply:
OK, I won't be able to try that until next week, but thanks, it's clear now

Reply:
FWIW, I believe there will be a ton of work before this API turns into something reasonable. I'm not even sure it's doable in the 4.0 release timeframe unless we want to release a half-baked solution.